src/Components/GEOSldas_GridComp/GEOSlandassim_GridComp/clsm_ensupd_read_obs.F90

```diff
@@ -83,9 +83,439 @@ module clsm_ensupd_read_obs
+ real, dimension(:), pointer :: lon_c, lat_c
+ real, dimension(:), pointer :: MODIS_SCA
+
+ type(date_time_type),intent(in) :: date_time ! loading the UTC hour information to constrain the longitude of MODIS obs.
+ character(2):: HH
+
+ !local parameters
+ integer, parameter:: N_fields = 4 
+ character(30), dimension(N_fields), parameter:: field_names = (/ &
+  'Day_CMG_Snow_Cover',        &   !1
+  'Day_CMG_Clear_Index',       &   !2
+  'Day_CMG_Cloud_Obscured',    &   !3
+  'Snow_Spatial_QA'/)   !4
+
+ integer, parameter :: nodata = -9999 !note: integer
+
+ ! Quality control step
+ ! step 1: remove the Snow cover > 100 (e.g., lake, night, inland water, ocean..)
+ ! step 2: Clear Index (CI): CI > 20% is considered
+ !
+
+ !declariations of hdf functions
+ integer:: hopen, vfstart, vsfatch, vsqfnelt, vsfseek, vsfsfld, vsfread
+ integer:: vsfdtch, vfend, hclose
+
+ ! declarations of hdf-related parameters and variables
+ integer, dimension(N_files) :: file_id, vdata_id 
+ integer :: status, n_read, n_records, record_pos
+ integer, parameter :: num_dds_block = 0  ! only important for writing hdf
+ integer, parameter :: vdata_ref = 4
+ integer, parameter :: DFACC_READ     = 1 ! from hdf.inc
+ integer, parameter :: FULL_INTERLACE = 0 ! from hdf.inc
+ 
+ ! local variables
+ logical :: must_stop
+ integer, dimension(N_files) :: N_data_tmp
+
+ ! variables to define latitude and longitude    
+ integer :: i, j, k, k_off, ll, mm, kk
+ integer :: hh_num, time_index
+
+ integer, parameter :: bin_size = 0.05
+ integer, parameter :: XGRID = 3600
+ integer, parameter :: YGRID = 7200
+
+ real :: lat_ind(XGRID) = (/(ll, ll=0, 3600-1, 1)/)
+ real :: lon_ind(YGRID) = (/(mm, mm=0, 7200-1, 1)/)
+
+ !-------------------------------------------------------------
+ ! UTC | 0000    | 0300   | 0600  | 0900 | 1200 | 1500   | 1800    | 2100    
+ ! lon | 180 135 | 135 90 | 90 45 | 45 0 | 0 -45| -45 -90| -90 -135| -135 -180
+ !
+
+ real:: lon_subtime(9) = (/(kk, kk=180, -180, -45)/)
+     
+ integer, dimension(:), allocatable :: Snow_QA
+ integer, dimension(:), allocatable :: CI_Index, Cloud_index
+  
+ integer*2, dimension(:), allocatable :: tmpint2vec
+ real,      dimension(:), allocatable :: tmprealvec
+
+ character(len=*), parameter :: Iam = 'read_MOD10C1_hdf'
+ character(len=400) :: err_msg
+
+ lat_c = (90-bin_size/2)-bin_size*lat_ind
+ lon_c = (-180+bin_size/2)-bin_size*lon_ind
+
+ do i=1,7200
+ lat(3600*(i-1)+1:3600*i)= lat_c
+ lon(3600*(i-1)+1:3600*i)= lon_c(i)
+ end do
+
+  ! determine number of data to be read from each file
+    
+    do j=1,N_files
+       
+       ! open and "start" hdf file 
+       file_id(j) = hopen( fnames(j), DFACC_READ, num_dds_block )  
+       status = vfstart(file_id(j))
+         
+       ! select vdata block that contains fields of interest
+       vdata_id(j) = vsfatch(file_id(j), vdata_ref, 'r') 
+       
+       ! determine number of records in vdata
+       status = vsqfnelt(vdata_id(j), n_records)
```

Comment:
I'm guessing the template for this subroutine was inherited from a reader for swath-based AMSR-E data.  In the case of the swath-based AMSR-E data, each file might have contained a different number of obs.  Here, however, each daily MODIS file contains a full global grid of 3600-by-7200 data points.  While querying "n_records" isn't wrong, it doesn't seem necessary.   That is, N_data_tmp could probably just be hard-wired as 3600*7200.